File: Data Structures & Algorithms/guess-number-higher-or-lower/submission-4.ts
/**
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * function guess(num) {}
 */

class Solution {
    /**
     * @param {number} n
     * @return {number}
     */
    guessNumber(n: number): number {
        let lo = 1;
        let hi = n;

        while (lo <= hi) {
            const mid = Math.floor((lo + hi) / 2);
            const res = guess(mid);
            if (res < 0) {
                hi = mid - 1;
            } else if (res > 0) {
                lo = mid + 1;
            } else {
                return mid;
            }
        }
    }
}

File: Data Structures & Algorithms/is-palindrome/submission-4.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        L = 0
        R = len(s) - 1
        s = s.lower()

        while L < R:
            while L < len(s) and not s[L].isalnum():
                L += 1
            
            while R >= 0 and not s[R].isalnum():
                R -= 1

            if L < R and s[L] != s[R]:
                print(f'{s[L]} {s[R]}')
                return False
            
            L += 1
            R -= 1

        return True
        